pallets/pallet-bonded-coins/src/tests/transactions/burn_into.rs

```diff
@@ -1 +1,726 @@
+use frame_support::{
+	assert_err, assert_ok,
+	traits::{
+		fungibles::Inspect,
+		tokens::{Fortitude, Preservation},
+	},
+};
+use frame_system::{pallet_prelude::OriginFor, RawOrigin};
+use sp_core::bounded_vec;
+use sp_runtime::{assert_eq_error_rate, traits::Scale, Permill, TokenError};
 
+use crate::{
+	mock::{runtime::*, *},
+	types::{Locks, PoolStatus},
+	Error, PoolDetailsOf,
+};
+
+// should not be u128::MAX, as a bug in the assets pallet results in transfers
+// failing if amount + total supply > u128::MAX
+const LARGE_BALANCE: u128 = u128::MAX / 10;
+
+const MAX_ERROR: Permill = Permill::from_perthousand(1);
+
+#[test]
+fn burn_first_coin() {
+	let pool_id = calculate_pool_id(&[DEFAULT_BONDED_CURRENCY_ID]);
+
+	let amount_to_burn: u128 = 1;
+	let expected_price =
+		(2 * amount_to_burn.pow(2) + 3 * amount_to_burn) * 10u128.pow(DEFAULT_COLLATERAL_DENOMINATION.into());
+
+	ExtBuilder::default()
+		.with_native_balances(vec![(ACCOUNT_00, ONE_HUNDRED_KILT)])
+		.with_collaterals(vec![DEFAULT_COLLATERAL_CURRENCY_ID])
+		.with_bonded_balance(vec![
+			(DEFAULT_COLLATERAL_CURRENCY_ID, pool_id.clone(), LARGE_BALANCE),
+			(DEFAULT_BONDED_CURRENCY_ID, ACCOUNT_00, amount_to_burn),
+		])
+		.with_pools(vec![(
+			pool_id.clone(),
+			PoolDetailsOf::<Test> {
+				curve: get_linear_bonding_curve(),
+				manager: None,
+				transferable: true,
+				bonded_currencies: bounded_vec![DEFAULT_BONDED_CURRENCY_ID],
+				state: PoolStatus::Active,
+				collateral_id: DEFAULT_COLLATERAL_CURRENCY_ID,
+				denomination: 0,
+				owner: ACCOUNT_99,
+			},
+		)])
+		.build()
+		.execute_with(|| {
+			let origin = RawOrigin::Signed(ACCOUNT_00).into();
+
+			assert_ok!(BondingPallet::burn_into(
+				origin,
+				pool_id,
+				0,
+				ACCOUNT_00,
+				amount_to_burn,
+				expected_price - 1, // rounding down may be happening in the conversion to fixed
+				1
+			));
+
+			assert_eq_error_rate!(
```

Comment:
Thats nice :grimacing: I would prob use the same macro for the curves tests.

Reply:
it does not seem to work for the Fixed type for some reason :(

Reply:
Oh, okay. That's unfortunate.